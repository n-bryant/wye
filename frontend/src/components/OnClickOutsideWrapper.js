import React, { Component } from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core";
import createClassNameGenerator from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";

// supply a root classname here
const styles = theme => ({
  root: {}
});

/**
 * Wrapper component for handling on click outside interactions with the provided children
 */
export class OnClickOutsideWrapper extends Component {
  componentDidMount = () => {
    document.addEventListener("click", this.checkIfOutside);
  };

  componentWillUnmount = () => {
    document.removeEventListener("click", this.checkIfOutside);
  };

  /**
   * Set the wrapper ref
   */
  setWrapperRef = node => {
    this.wrapperRef = node;
  };

  /**
   * Check if a click has occurred outside of the children and call the passed handler if it has
   */
  checkIfOutside = e => {
    if (this.wrapperRef && !this.wrapperRef.contains(e.target)) {
      this.props.handleClickOutside();
    }
  };

  render() {
    const { classes } = this.props;
    const classnames = OnClickOutsideWrapper.classnames({ classes });
    return (
      <div ref={this.setWrapperRef} className={classnames.root()}>
        {this.props.children}
      </div>
    );
  }
}
OnClickOutsideWrapper.classnames = createClassNameGenerator(
  `${JSS_CLASS_NAME_PREFIX}OnClickOutsideWrapper`
);
OnClickOutsideWrapper.propTypes = {
  // root classname
  classes: PropTypes.shape({
    root: PropTypes.string
  }),
  // what to render inside the wrapper
  children: PropTypes.element.isRequired,
  // the function to call when a click occurrs outside the provided children
  handleClickOutside: PropTypes.func.isRequired
};

export default withStyles(styles)(OnClickOutsideWrapper);
