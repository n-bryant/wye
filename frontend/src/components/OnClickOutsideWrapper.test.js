import { OnClickOutsideWrapper } from "./OnClickOutsideWrapper";

const props = {
  classes: {
    root: "root"
  },
  children: <p>foo</p>,
  handleClickOutside: jest.fn()
};
describe("OnClickOutsideWrapper", () => {
  it("should render a div containing the children provided", () => {
    expect(
      toJson(shallow(<OnClickOutsideWrapper {...props} />))
    ).toMatchSnapshot();
  });

  it("should set a ref on the root element", () => {
    expect(
      mount(<OnClickOutsideWrapper {...props} />).instance().wrapperRef
    ).toBeDefined();
  });

  it("should call the handleClickOutside prop when something is clicked outside the children", () => {
    const wrapper = mount(<OnClickOutsideWrapper {...props} />, {
      attachTo: document.body
    });
    const event = {
      target: document.createElement("div")
    };
    wrapper
      .find(OnClickOutsideWrapper)
      .instance()
      .checkIfOutside(event);
    expect(props.handleClickOutside).toHaveBeenCalled();
  });
});
