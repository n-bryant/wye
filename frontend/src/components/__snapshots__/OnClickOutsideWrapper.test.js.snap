// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnClickOutsideWrapper should render a div containing the children provided 1`] = `
<div
  className="wye-OnClickOutsideWrapper root"
>
  <p>
    foo
  </p>
</div>
`;
